File: design_patterns/c++/src/create/abstract_factory/MazeFactory.cpp
#include <cstdlib>
#include <cstring>
#include "../../../headers/create/abstract_factory/MazeFactory.h"
#include "../../../headers/create/abstract_factory/BombedMazeFactory.h"
#include "../../../headers/create/abstract_factory/EnchantedMazeFactory.h"

Maze* MazeFactory::makeMaze() const {
    return new Maze;
}
Wall* MazeFactory::makeWall() const {
    return new Wall;
}
Room* MazeFactory::makeRoom(int number) const {
    return new Room(number);
}

Door* MazeFactory::makeDoor(Room* r1, Room* r2) {
    return new Door(r1, r2);
}

File: design_patterns/java/src/create/abstract_factory/MazeFactory.java
package create.abstract_factory;/*
* The MazeFactory not only as to abstract factroy but also
* as a concret factory. So it is not a abstract class. It
* can reduce the number of classes.
*/

import base.Door;
import base.Maze;
import base.Room;
import base.Wall;

public class MazeFactory {
    public Maze makeMaze() {
        return new Maze();
    }

    public Wall makeWall() {
        return new Wall();
    }

    public Room makeRoom(int num) {
        return new Room(num);
    }

    public Door makeDoor(Room r1, Room r2) {
        return new Door(r1, r2);
    }

}
